### DISCUSSION TITLE
Set zfs_abd_scatter_enabled = 0 on 64bit Kernel?

### DISCUSSION THREAD

**Discussion by jxdking** [openzfs/zfs#11944] [Q&A] [answered]
From what I understand, by given large enough recordsize, when it is enabled, the data is split into chunks and allocated on kernel logical address space. When it is disabled, the data is allocated through vmalloc.
It is just a software implementation of what vmalloc does (stitching scattered physical pages into one), while virtual address mapping is hardware accelerated. Virtual address space also allows memory compaction if free memory is highly fragmented.
What makes scattered abd worse is that it cannot be consumed directly, it has to be copy to a regular buffer before read and write, which doubles memory usage in some cases.
Since 64bit kernel has enough virtual address space, why not disable zfs_abd_scatter_enabled. 

 Did I miss something?

**Comment by amotin** [comment, marked as answer]
While KVA is not so strictl constrained on 64-bit architectures, AFAIK it is still not free.  Also mapping/unmapping the KVA is expensive process, requiring global TLB shootdowns, that is also expensive.
